pkg/controller/infra/infra_controller.go

```diff
@@ -118,99 +138,369 @@ type InfraController struct {
+
+	return true
 }
 
-func (c *InfraController) processNextWorkItem() bool {
-	key, quit := c.queue.Get()
+func (c *InfraController) processNextJobWorkItem() bool {
+	key, quit := c.jobQueue.Get()
 	if quit {
 		return false
 	}
-	defer c.queue.Done(key)
+	defer c.jobQueue.Done(key)
 
-	err := c.syncHandler(key.(string))
-	c.handleErr(err, key)
+	err := c.jobSyncHandler(key.(string))
+	c.handleJobErr(err, key)
 
 	return true
 }
 
-func (c *InfraController) handleErr(err error, key interface{}) {
+func (c *InfraController) handleClusterErr(err error, key interface{}) {
 	if err == nil {
-		c.queue.Forget(key)
+		c.clusterQueue.Forget(key)
 		return
 	}
 
-	c.logger.WithField("cluster", key).Errorf("error syncing cluster: %v", err)
-	if c.queue.NumRequeues(key) < maxRetries {
-		c.logger.WithField("cluster", key).Errorf("error syncing cluster: %v", err)
-		c.queue.AddRateLimited(key)
+	logger := c.logger.WithField("cluster", key)
+
+	logger.Errorf("error syncing cluster: %v", err)
+	if c.clusterQueue.NumRequeues(key) < maxRetries {
+		logger.Errorf("retrying cluster")
+		c.clusterQueue.AddRateLimited(key)
+		return
+	}
+
+	utilruntime.HandleError(err)
+	logger.Infof("dropping cluster out of the queue: %v", err)
+	c.clusterQueue.Forget(key)
+}
+
+func (c *InfraController) handleJobErr(err error, key interface{}) {
+	if err == nil {
+		c.jobQueue.Forget(key)
+		return
+	}
+
+	logger := c.logger.WithField("job", key)
+
+	logger.Errorf("error syncing job: %v", err)
+	if c.jobQueue.NumRequeues(key) < maxRetries {
+		logger.Errorf("retrying job")
+		c.jobQueue.AddRateLimited(key)
 		return
 	}
 
 	utilruntime.HandleError(err)
-	c.logger.WithField("cluster", key).Infof("dropping cluster out of the queue: %v", err)
-	c.queue.Forget(key)
+	logger.Infof("dropping job out of the queue: %v", err)
+	c.jobQueue.Forget(key)
+}
+
+// syncJob will sync the job with the given key.
+// This function is not meant to be invoked concurrently with the same key.
+func (c *InfraController) syncJob(key string) error {
+	startTime := time.Now()
+	cLog := c.logger.WithField("job", key)
+	cLog.Debugln("started syncing job")
+	defer func() {
+		cLog.WithField("duration", time.Since(startTime)).Debugln("finished syncing job")
+	}()
+
+	ns, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return err
+	}
+	if len(ns) == 0 || len(name) == 0 {
+		return fmt.Errorf("invalid job key %q: either namespace or name is missing", key)
+	}
+
+	job, err := c.jobsLister.Jobs(ns).Get(name)
+	if errors.IsNotFound(err) {
+		cLog.Warnln("job not found")
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+
+	cluster, err := c.clusterForJob(job)
+	if err != nil {
+		return fmt.Errorf("could not retrieve cluster for job %s/%s: %v", job.Namespace, job.Name, err)
+	}
+	if cluster == nil {
+		return nil
+	}
+
+	// only proceed if this is the current provisioning job for the cluster
+	if cluster.Status.ProvisioningJob == nil || cluster.Status.ProvisioningJob.Name != job.Name {
```

Comment:
I think that we have a race condition here? There is a real possibility that the controller will sync the provision job before getting the update cluster object with its `ProvisionJob` field set. This will cause the controller to skip the sync of the job. It is even possible (although not practically) that the job finished completely before the lister gets the update cluster object. In that worst case scenario, the controller will never sync the job.